kernel/src/schema_compat.rs

```diff
@@ -0,0 +1,331 @@
+//!  # use delta_kernel::schema::DataType;
+//!  let schema = StructType::new([
+//!     StructField::new("id", DataType::LONG, false),
+//!     StructField::new("value", DataType::STRING, true),
+//!  ]);
+//!  let read_schema = StructType::new([
+//!     StructField::new("id", DataType::LONG, true),
+//!     StructField::new("value", DataType::STRING, true),
+//!     StructField::new("year", DataType::INTEGER, true),
+//!  ]);
+//!  assert!(schema.can_read_as(&read_schema).is_ok());
+//!  ````
+//!
+//! [`Schema`]: crate::schema::Schema
+use std::collections::{HashMap, HashSet};
+
+use crate::utils::require;
+use crate::{DeltaResult, Error};
+
+/// The nullability flag of a schema's field. This can be compared with a read schema field's
+/// nullability flag using [`NullabilityFlag::can_read_as`].
+struct NullabilityFlag(bool);
+
+impl NullabilityFlag {
+    /// Represents a nullability comparison between two schemas' fields. Returns true if the
+    /// read nullability is the same or wider than the nullability of self.
+    fn can_read_as(&self, read_nullable: NullabilityFlag) -> DeltaResult<()> {
+        // The case to avoid is when the column is nullable, but the read schema specifies the
+        // column as non-nullable. So we avoid the case where !read_nullable && nullable
+        // Hence we check that !(!read_nullable && existing_nullable)
+        // == read_nullable || !existing_nullable
+        require!(
+            read_nullable.0 || !self.0,
+            Error::generic("Read field is non-nullable while this field is nullable")
+        );
+        Ok(())
+    }
+}
+
+impl crate::schema::StructField {
+    /// Returns `Ok` if this [`StructField`] can be read as `read_field` in the read schema.
+    fn can_read_as(&self, read_field: &Self) -> DeltaResult<()> {
+        NullabilityFlag(self.nullable).can_read_as(NullabilityFlag(read_field.nullable))?;
+        require!(
+            self.name() == read_field.name(),
+            Error::generic(format!(
+                "Struct field with name {} cannot be read with name {}",
+                self.name(),
+                read_field.name()
+            ))
+        );
+        self.data_type().can_read_as(read_field.data_type())?;
+        Ok(())
+    }
+}
+impl crate::schema::StructType {
+    /// Returns `Ok` if this [`StructType`] can be read as `read_type` in the read schema.
+    #[allow(unused)]
+    pub(crate) fn can_read_as(&self, read_type: &Self) -> DeltaResult<()> {
+        let field_map: HashMap<String, &crate::schema::StructField> = self
+            .fields
+            .iter()
+            .map(|(name, field)| (name.to_lowercase(), field))
+            .collect();
+        require!(
+            field_map.len() == self.fields.len(),
+            Error::generic("Delta tables don't allow field names that only differ by case")
+        );
+
+        let read_field_names: HashSet<String> =
+            read_type.fields.keys().map(|x| x.to_lowercase()).collect();
```

Comment:
for this and usage below: perhaps for now we can specify this is a lowercase hashset? but in the future I hope we can do something better to enforce the case-insensitivity instead of relying on `to_lowercase` before interaction with the set?

Reply:
Changed the name to reflect that  In terms of future, Idk how else to do this check. Delta-spark uses a case-insensitive map which I don't think we should add to kernel.  Also for clarity, the lowercase is being used for two reasons:
1) To ensure that the schema is valid by asserting that it doesn't have duplicate fields that differ by case
2) To find fields shared between this schema and the read schema. Note that we later assert that the field names are exactly the same.